standard/variables.md

```diff
@@ -901,3 +901,181 @@ variable_reference
+
+### §ref-span-safety-escape-scopes Safe scopes
+
+For any variable, the *ref_safe_scope* of that variable is the scope where a *variable_reference* (§9.5) to that variable is valid. The *referent* of a reference variable must have a *ref_safe_scope* that is at least as wide as the scope where the reference variable is valid.
+
+There are three valid *ref_safe_scope*s:
+
+- *block*: A variable declared in a block is valid in that block in which it is declared. This includes all local variables. A local variable declared in a method is valid in the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block, or a nested block.
+- *method*: A value parameter on a method declaration, including the implicit `this` parameter is valid in the entire method. The fields of a `struct` type are valid in the entire method. A variable with *ref_safe_scope* of *method* is a valid referent only if the reference variable is declared in the same method.
+- *caller_scope*: Member fields of a `class` type and `ref` parameters have *ref_safe_scope* of *caller_scope*. A variable with *ref_safe_scope* of *caller_scope* can be the referent of a reference return. A variable that can be referent of a *reference-return* is *safe-to-ref-return*.
+
+These values form a nesting relationship from narrowest (*block*) to widest (*caller-scope*).
+
+For any variable, the *safe_scope* of that variable is the scope where its value may be copied. A variable whose type is not a `ref struct` type is *safe-to-return* from the entire enclosing method. Its *safe_scope* is *caller_method*. Otherwise the following rules apply.
+
+> *Example*: The following code shows examples of the different *safe_scope* and *ref_safe_scope* values. The declarations show values of *safe_scope* for a referent to be the initializing expression for a `ref` variable. The examples show the values of *ref_safe_scope* for a reference return:
+>
+> ```csharp
+> public class C
+> {
+>     private int[] arr = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+> 
+>     public ref int M1(ref int rv)
+>     {
+>         return ref rv; // rv is safe to ref return
+>     }
+>
+>     public ref int M2(int v)
+>     {
+>         return ref v; // error: v isn't safe to ref return
+>     }
+>
+>     public ref int M3(ref int rv)
+>     {
+>         int v = 5;
+> 
+>         return ref arr[v]; // arr[v] is safe to ref return
+>     }
+> 
+>     public void M4(int p) 
+>     {
+>         int v = 6;
+>         ref int rv = ref p; // safe scope of rv is block, ref safe scope of p is method
+>         ref int rv2 = ref v; // safe scope of rv2 is block, ref safe scope of v is block
+>         ref int rv3 = ref arr[v]; // safe scope of rv3 is block, ref safe scope of arr[v] is caller method
+>     }
+> }
+> ```
+>
+> *end example.*
+
+### §ref-span-safety-locals Local variable safe scope
+
+For a local variable `v`:
```

Comment:
I was surprised to see the rule for determining a ref_safe_scope in a section about safe scopes

Reply:
Aha.  This gives me an idea:

The concept of *safe_scope* only is interesting with respect to `ref struct` types. For all other types, you can obviously pass by value and return by value. A `ref struct` is interesting ***because*** it's *safe_scope* is constrained by the *ref_safe_scope* of any member *ref-like* fields.

After I make the other edits. I'd like your thoughts on the following:

- Don't introduce *safe_scope* as a term.
- Instead, add a rule that any instance of a `ref struct` is constrained to the *ref_safe_scope* of the referent in any *ref_like_fields*.

Reply:
Using one term should work just fine as long as we are consistent. Considered doing this once or twice in the ref safety docs when we implemented `ref` fields but was too much work to go back and change the original spec.

Slight tangent ... if you think about `ref` as a type vs. an annotation on a location it often makes the rules around ref safety easier to grok (at least to me). Consider that if we had `ref<int>` instead of `ref int` then having a single term becomes completely natural here as everything is just a value.

Reply:
Thanks @jaredpar - that's really interesting. On my next pass through this, I'll try reading it from that perspective. @BillWagner if that turns out to be positive, *maybe* we could add a note? It's not the kind of note we'd often add, but given the complexities of this, it might be worth it here.